### ISSUE TITLE
Camelized Named arguments naming issue

### ISSUE THREAD

**Issue by lifeart** [emberjs/ember.js#17377]
`<@ItemComponent />` renders "OK", `{{@ItemComponent}}` -> throw error `Assertion Failed: '@ItemComponent' is reserved.`
reproduction: [Ember Twiddle](https://ember-twiddle.com/b4f0d498611309dbcb26c9577b768f17?openFiles=templates.components.my-books.books-list.hbs%2C)

Refs: https://emberjs.github.io/rfcs/0276-named-args.html#reserved-names

It's really confusing if you missed this paragrath in RFC, should we update error message to point to RFC link or offer differen name? Instead of Assertion Failed: '@ItemComponent' is reserved. show something like this: Assertion Failed: '@ItemComponent' is reserved. RFC 276#reserved-names try change {componentName} attribute name from "ItemComponent" to "itemComponent".

**Comment by pixelhandler** [comment]
@lifeart I see how that using `{{@ItemComponent}} like an argument value when `ItemComponent` is a component should error. And that the error mentions that the variable is reserved. Perhaps it should just say "Cannot render X argument, did you intend to invoke a component? Change @X to X". It would be interesting if the template compiler did recognized the mistake. I'm not sure what should be the proper error.

**Comment by rwjblue** [comment]
Yes, the error message should at least be fleshed out a bit more. We should probably also at least _consider_ if we really do want to reserve all initial capitalized named args (we originally did this speculatively).

**Comment by chancancode** [comment]
I'm 👍 on linking to the RFC in the error message, on debug builds at least. I'm also open to revisiting the reserved names, it was definitely broader than it needed to be, but at the same time I'm not sure how else to predict what we may need here.

**Comment by chancancode** [comment]
By the way, should we at least make the passing side match the receiving side?

**Comment by locks** [comment]
Is this still relevant?

**Comment by amk221** [comment]
I'm experiencing this, with the following scenario:

```handlebars
<Foo @Bar={{@Bar}} />
```

...which I have to work around like:

```javascript
get bar() {
  return this.args.Bar;
}
```

```handlebars
<Foo @Bar={{this.bar}} />
```

**Comment by pzuraq** [comment]
This seems like something we may want to revisit in terms of reserving arg names, I think capital cased names are much going to make even more sense now and in the near future. @amk221 maybe consider opening an RFC to un-reserve these names?

And I agree, we should have a better error message here either way. Explanations of what is reserved in general.

**Comment by lifeart** [comment]
Related linting rule https://github.com/ember-template-lint/ember-template-lint/pull/1759

**Comment by amk221** [comment]
I'd also like to bring up `@_` which is also reserved, it's natural to reach for an underscored argument, especially when building closure components.

<details>
  <summary>example</summary>

```javascript
// application.js
export default class extends Controller {
  @action
  handleChosingAThing() {
    // Main public usage
  }
}
```
```javascript
// foo.js
export default class extends Component {
  @action
  handleChosingAThing() {
    // Internal usage of handling a thing
  }
}
```
```handlebars
{{! foo.hbs }}
{{yield (hash Bar=(component "bar" _onChoseAThing=this.handleChosingAThing))}}
```
```handlebars
{{! bar.hbs }}
<button {{on "click" @onChoseAThing}} {{on "click" @_onChoseAThing}}></button>
```
```handlebars
{{! application.hbs }}
<Foo as |foo|>
  <foo.Bar @onChoseAThing={{this.choseAThing}} />
</Foo>
```
</details>

**Comment by bertdeblock** [comment]
I made https://github.com/emberjs/rfcs/issues/803 as a first step for this. Feedback appreciated!
